text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+  'arn:aws:s3:::mybucketname'
+);
+
+const oac = new cloudfront.OriginAccessControl(this, 'MyOAC', {
+  originAccessControlOriginType: cloudfront.OriginAccessControlOriginType.S3,
+});
+
+const distribution = new cloudfront.Distribution(this, 'MyDistribution', {
+  defaultBehavior: {
+    origin: new origins.S3Origin(bucket, {
+      originAccessControl: oac,
+      overrideImportedBucketPolicy: true
+    })
+  }
+});
+```
+
+# CloudFront Origins for the CDK CloudFront Library
+
+## S3 Bucket
+
+An S3 bucket can be added as an origin. If the bucket is configured as a website endpoint, the distribution can use S3 redirects and S3 custom error
+documents.
+
+```ts
+const myBucket = new s3.Bucket(this, 'myBucket');
+new cloudfront.Distribution(this, 'myDist', {
+  defaultBehavior: { origin: new origins.S3Origin(myBucket) },
+});
+```
+
+The above will treat the bucket differently based on if `IBucket.isWebsite` is set or not. If the bucket is configured as a website, the bucket is
+treated as an HTTP origin, and the built-in S3 redirects and error pages can be used. Otherwise, the bucket is handled as a bucket origin and
+CloudFront's redirect and error handling will be used.
+
+### Restricting access to an S3 Origin
+
+CloudFront provides two ways to send authenticated requests to an Amazon S3 origin: origin access control (OAC) and origin access identity (OAI).
+OAC is the recommended method and OAI is considered legacy (see [Restricting access to an Amazon Simple Storage Service origin](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html)).
+Following AWS best practices, it is recommended you set the feature flag `@aws-cdk/aws-cloudfront:useOriginAccessControlByDefault` to `true` to use OAC by
+default when creating new origins.
+
+For an S3 bucket that is configured as a standard S3 bucket origin (not as a website endpoint), when the above feature flag is enabled the `S3Origin`
+construct will automatically create an OAC and grant it access to the underlying bucket.
+
+> [Note](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html): When you use OAC with S3
+bucket origins you must set the bucket's object ownership to Bucket owner enforced, or Bucket owner preferred (only if you require ACLs).
+
+```ts
+const myBucket = new s3.Bucket(this, 'myBucket', {
+  objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_ENFORCED,
+});
+new cloudfront.Distribution(this, 'myDist', {
+  defaultBehavior: { 
+    origin: new origins.S3Origin(myBucket) // Automatically creates an OAC
+  },
+});
+```
+
+Alternatively, a custom origin access control can be passed to the S3 origin:
+
+```ts
+const myBucket = new s3.Bucket(this, 'myBucket');
+const myOAC = new cloudfront.OriginAccessControl(this, 'myOAC', {
+  description: 'Origin access control for S3 origin',
+  originAccessControlOriginType: cloudfront.OriginAccessControlOriginType.S3,
```

Comment:
Discussed offline, but still worth mentioning; `Type` properties on constructs should not be used. These cause a bad user experience due to runtime error checking for conflicting props, like [this](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-ecs/lib/fargate/fargate-service.ts#L124). These errors become hard for us to maintain, but more importantly they negatively impact the user experience because the in-IDE code completion suggestions lead them to incorrect configuration. The autocomplete should guide them only to error-free prop combinations, and runtime error checking negates that. 

We can solve this by hiding the Type property entirely and creating a base class with multiple subclasses. So here you'll want a `S3OriginAccessControlOrigin` class that's visible to users, and a not-visible base class for `OriginAccessControlOrigin`. Only the props available to that particular class should be exposed on its props, and the props applicable to all should be on the base class. Props available to most, but not all of the subclasses, should be added to each subclass, unless it gets egregious. See the [Batch constructs](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-batch/lib/managed-compute-environment.ts) for an example of this pattern, particularly `ManagedComputeEnvironment` and `JobDefinition`.

Reply:
Makes sense. In this case the only difference between the subclasses is the `originAccessControlOriginType` property and the other properties are the same, so I opted to add a public static method `forS3()` which sets the type property accordingly (need to update the example here).  Creating subclasses means each subclasses's props would just include the `originAccessControlOriginType` property. But we could then make the base class abstract so users can only create instances of the subclass depending on the origin type, which better accomplishes what you mentioned about autocomplete guiding them to the correct configuration. I'll update the example.

Reply:
I'd need to see the example but I think y'all are on the right path here.

Reply:
The base class should be not only abstract but also `@internal`, which will prevent it from being exposed to users. 

The subclass props do not need to include the `Type` property. The implementing class should configure that transparently to the user. 

It may be true today that there is minimal or no difference between these types, but it almost certainly will not be true in the future. Every other AWS resource I've encountered that has a `Type` property has conflicting properties, and it only gets ever more complex over time. 

For example AWS Batch used to support one type of container. The old constructs for that resource made the only supported container type a required property, and when the resource added more types in the future, the construct could not support them. We should always assume that Type fields imply hidden property conflicts that you either haven't discovered or haven't been implemented by the service team yet.

Reply:
Updated to create an abstract base class for `OriginAccessControl` with subclasses for each origin type